File: sorting/selection_sort.rb
# Keeps a sorted and unsorted portion of array. Starting with first value, SELECTs the min value when traversing over unsorted part of array. Then swaps that value with the leftmost unsorted value (or current value)

def selectionSort(input)
  input.each_with_index do |num1, i|
    min = num1
    min_index = i
    ((i+1)..(input.length - 1)).each do |j|
      p "j #{j}"
      if input[j] < min
        min = input[j]
        min_index = j
      end
    end
    input[min_index] = input[i]
    input[i] = min
  end
  input
end

class SelectionSortTest < Test::Unit::TestCase
  def test_selectionSort_should_handle_example_case
    test = selectionSort([8,3,2,10])
    answer = [2,3,8,10]

    assert_equal(answer, test);
  end

  def test_selectionSort_should_handle_empty_input
    test = selectionSort([])
    answer = []

    assert_equal(answer, test);
  end

  def test_selectionSort_should_handle_large_input
    testInput = []
    $i = 1000000

    while $i > 0
      toPush = Random.rand(1000000)
      testInput.push(toPush)
      $i = $i - 1
    end

    test = selectionSort(testInput)
    assert_equal(testInput.sort, test)
  end

end

File: sorting/selection_sort.js
var selectionSort = function(input){
  for (var i = 0; i < input.length; i++) {
    var min = input[i];
    var minIndex = i;
    console.log('min', min, 'minIndex', minIndex)
    for (var j = i+1;j < input.length; j++) {
      console.log('j',j)
      if (input[j] < min) {
        min = input[j];
        minIndex = j;
      }
    }
    input[minIndex] = input[i];
    input[i] = min;
  }
  return input;
}


// describe('SELECTION SORT ', function(){

//   describe("RUN ON [8, 3, 2, 10]: ", function(){
//     it("should return [2, 3, 8, 10]", function(){
//       var test = selectionSort([8,3,2,10]);
//       var answer = [2,3,8,10];

//       expect(test).to.eql(answer);
//     });
//   });

//   describe("RUN ON EMPTY INPUT []: ", function(){
//     it("should return []", function(){
//       var test = selectionSort([]);
//       var answer = [];

//       expect(test).to.eql(answer);
//     });
//   });

//   describe("RUN ON LARGE INPUT: ", function(){
//     it("should complete", function(){
//       var testInput = [];
//       var i = 1000000;
//       while (i--){
//         testInput.push(Math.floor(Math.random() * 1000000))
//       }
//       var test = selectionSort(testInput);

//       expect(testInput.sort(function(a, b){return a-b})).to.eql(test);
//     })
//   })

// });
